DynamicTablesPkg/Library/Acpi/Common/AcpiDbg2Lib/Dbg2Generator.c

```diff
@@ -356,161 +394,318 @@ BuildDbg2TableEx (
-    // Default to DWORD access size as the access
-    // size field was introduced at a later date
-    // and some ConfigurationManager implementations
-    // may not be providing this field data
-    AcpiDbg2.Dbg2DeviceInfo[INDEX_DBG_PORT0].BaseAddressRegister.AccessSize =
-      EFI_ACPI_6_3_DWORD;
-  } else {
-    AcpiDbg2.Dbg2DeviceInfo[INDEX_DBG_PORT0].BaseAddressRegister.AccessSize =
-      SerialPortInfo->AccessSize;
-  }
+  CurrentDbg2Device = (EFI_ACPI_DBG2_DEBUG_DEVICE_INFORMATION_STRUCT *)((UINT8 *)AcpiDbg2 + AcpiDbg2->OffsetDbgDeviceInfo);
 
-  // Update the serial port subtype
-  AcpiDbg2.Dbg2DeviceInfo[INDEX_DBG_PORT0].Dbg2Device.PortSubtype =
-    SerialPortInfo->PortSubtype;
+  // Populate DBG2 devices
+  for (Index = 0; Index < Dbg2DeviceCount; Index++) {
+    if ((CurrentDbg2Device == NULL) ||
+        ((UINTN)CurrentDbg2Device - (UINTN)AcpiDbg2 > AcpiDbg2Len))
+    {
+      Status = EFI_INVALID_PARAMETER;
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Device pointer out of range\n"));
+      goto error_handler;
+    }
 
-  if ((SerialPortInfo->PortSubtype ==
-       EFI_ACPI_DBG2_PORT_SUBTYPE_SERIAL_ARM_PL011_UART)           ||
-      (SerialPortInfo->PortSubtype ==
-       EFI_ACPI_DBG2_PORT_SUBTYPE_SERIAL_ARM_SBSA_GENERIC_UART_2X) ||
-      (SerialPortInfo->PortSubtype ==
-       EFI_ACPI_DBG2_PORT_SUBTYPE_SERIAL_ARM_SBSA_GENERIC_UART))
-  {
-    // Setup the PL011 length.
-    AcpiDbg2.Dbg2DeviceInfo[INDEX_DBG_PORT0].AddressSize = PL011_UART_LENGTH;
+    if ((Dbg2DeviceInfo == NULL) || (Dbg2DevicesMemoryRange == NULL) || (Dbg2DevicesMemoryRangeCount == NULL)) {
+      Status = EFI_INVALID_PARAMETER;
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Dbg2DeviceInfo, Dbg2DevicesMemoryRange, or Dbg2DevicesMemoryRangeCount is NULL. Status = %r\n", Status));
+      goto error_handler;
+    }
 
-    // Initialize the serial port
-    Status = SetupDebugUart (SerialPortInfo);
+    Status = PopulateDbg2Device (
+               &CurrentDbg2Device,
+               &Dbg2DeviceInfo[Index],
+               Dbg2DevicesMemoryRange[Index],
+               Dbg2DevicesMemoryRangeCount[Index]
+               );
     if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "ERROR: DBG2: Failed to populate device %u. Status = %r\n", Index, Status));
+      goto error_handler;
+    }
+  }
+
+  // Free these as they are not needed anymore
+  if (Dbg2DevicesMemoryRange != NULL) {
+    FreePool (Dbg2DevicesMemoryRange);
+    Dbg2DevicesMemoryRange = NULL;
+  }
+
+  if (Dbg2DevicesMemoryRangeCount != NULL) {
+    FreePool (Dbg2DevicesMemoryRangeCount);
+    Dbg2DevicesMemoryRangeCount = NULL;
+  }
+
+  // Currently only one serial port is supported
+  if (SerialPortCount > 0) {
+    Index                                     = 0; // Always use the first serial port
+    SerialMemoryRange.BaseAddress             = SerialPortInfo[Index].BaseAddress;
+    SerialMemoryRange.Length                  = SerialPortInfo[Index].BaseAddressLength;

```

Comment:
@samimujawar Just a note that now that the static table used to have the `CM_ARCH_COMMON_SERIAL_PORT_INFO.BaseAddressLength` field hard coded to `PL011_UART_LENGTH`. This allowed some `CM_ARCH_COMMON_SERIAL_PORT_INFO` to have a valid address length while not providing any value.

Now that we rely on the CM_OBJECT, some ConfigurationManager might need to be updated.

Not sure what is the best solution here: either add an error condition to prevent from generating a DBG2 table when no value is provided, or use `0x1000` by default if no value is provided ?